File: frontend/src/shared/lib/footer-phone.ts
export type FooterPhoneFormat = "mobile" | "toll_free_0800";

const FOOTER_PHONE_DIGITS = 10;

function normalizeFormat(value: string | null | undefined): FooterPhoneFormat {
  return value === "toll_free_0800" ? "toll_free_0800" : "mobile";
}

function normalizeMobileDigits(digits: string): string {
  if (!digits) {
    return "";
  }
  if (digits.startsWith("380")) {
    return digits.slice(2, 12);
  }
  if (digits.startsWith("38")) {
    const rest = digits.slice(2);
    if (!rest) {
      return "";
    }
    if (rest.startsWith("0")) {
      return rest.slice(0, FOOTER_PHONE_DIGITS);
    }
    return `0${rest}`.slice(0, FOOTER_PHONE_DIGITS);
  }
  if (digits.startsWith("0")) {
    return digits.slice(0, FOOTER_PHONE_DIGITS);
  }
  if (digits.length >= FOOTER_PHONE_DIGITS) {
    const tail = digits.slice(-FOOTER_PHONE_DIGITS);
    return tail.startsWith("0") ? tail : `0${tail}`.slice(0, FOOTER_PHONE_DIGITS);
  }
  return `0${digits}`.slice(0, FOOTER_PHONE_DIGITS);
}

function normalizeTollFreeDigits(digits: string): string {
  if (!digits) {
    return "";
  }
  if (digits.startsWith("380800")) {
    return `0${digits.slice(3)}`.slice(0, FOOTER_PHONE_DIGITS);
  }
  if (digits.startsWith("800")) {
    return `0${digits}`.slice(0, FOOTER_PHONE_DIGITS);
  }
  if (digits.startsWith("0800")) {
    return digits.slice(0, FOOTER_PHONE_DIGITS);
  }
  if (digits.length >= 6) {
    return `0800${digits.slice(-6)}`.slice(0, FOOTER_PHONE_DIGITS);
  }
  return `0800${digits}`.slice(0, FOOTER_PHONE_DIGITS);
}

export function normalizeFooterPhoneDigits(
  value: string,
  phoneFormat: FooterPhoneFormat = "mobile",
): string {
  const digits = String(value || "").replace(/\D+/g, "");
  if (!digits) {
    return "";
  }
  const normalizedFormat = normalizeFormat(phoneFormat);
  if (normalizedFormat === "toll_free_0800") {
    return normalizeTollFreeDigits(digits);
  }
  return normalizeMobileDigits(digits);
}

export function formatFooterPhoneForInput(
  value: string,
  phoneFormat: FooterPhoneFormat = "mobile",
): string {
  const normalizedFormat = normalizeFormat(phoneFormat);
  const normalized = normalizeFooterPhoneDigits(value, normalizedFormat);
  if (!normalized) {
    return "";
  }
  if (normalizedFormat === "toll_free_0800") {
    const code = normalized.slice(1, 4);
    const left = normalized.slice(4, 7);
    const right = normalized.slice(7, 10);
    let out = "0";
    if (code) {
      out += ` (${code}`;
      if (code.length === 3) {
        out += ")";
      }
    }
    if (left) {
      out += code.length === 3 ? ` ${left}` : left;
    }
    if (right) {
      out += `-${right}`;
    }
    return out;
  }

  const operator = normalized.slice(0, 3);
  const left = normalized.slice(3, 6);
  const middle = normalized.slice(6, 8);
  const right = normalized.slice(8, 10);

  let out = "38";
  if (operator) {
    out += ` (${operator}`;
    if (operator.length === 3) {
      out += ")";
    }
  }
  if (left) {
    out += operator.length === 3 ? ` ${left}` : left;
  }
  if (middle) {
    out += `-${middle}`;
  }
  if (right) {
    out += `-${right}`;
  }
  return out;
}

export function formatFooterPhoneForSave(
  value: string,
  phoneFormat: FooterPhoneFormat = "mobile",
): string {
  const normalizedFormat = normalizeFormat(phoneFormat);
  const normalized = normalizeFooterPhoneDigits(value, normalizedFormat);
  if (!normalized || normalized.length !== FOOTER_PHONE_DIGITS) {
    return "";
  }
  if (normalizedFormat === "toll_free_0800" && !normalized.startsWith("0800")) {
    return "";
  }
  return formatFooterPhoneForInput(normalized, normalizedFormat);
}

export function formatFooterPhoneDisplay(
  value: string,
  phoneFormat: FooterPhoneFormat = "mobile",
): string {
  const normalizedFormat = normalizeFormat(phoneFormat);
  const normalized = normalizeFooterPhoneDigits(value, normalizedFormat);
  if (normalized.length === FOOTER_PHONE_DIGITS) {
    if (normalizedFormat === "toll_free_0800" && !normalized.startsWith("0800")) {
      return String(value || "").trim();
    }
    return formatFooterPhoneForInput(normalized, normalizedFormat);
  }
  return String(value || "").trim();
}

File: backend/apps/marketing/services/footer_phone.py
from __future__ import annotations

import re

FOOTER_PHONE_DIGITS_COUNT = 10
FOOTER_PHONE_FORMAT_MOBILE = "mobile"
FOOTER_PHONE_FORMAT_TOLL_FREE_0800 = "toll_free_0800"
FOOTER_PHONE_FORMAT_EXAMPLE_MOBILE = "38 (0XX) XXX-XX-XX"
FOOTER_PHONE_FORMAT_EXAMPLE_0800 = "0 (800) XXX-XXX"


def normalize_footer_phone(value: str, phone_format: str = FOOTER_PHONE_FORMAT_MOBILE) -> str:
    normalized_format = _normalize_phone_format(phone_format)
    digits = _normalize_footer_phone_digits(value, normalized_format)
    if not digits:
        return ""
    if normalized_format == FOOTER_PHONE_FORMAT_TOLL_FREE_0800:
        if len(digits) != FOOTER_PHONE_DIGITS_COUNT or not digits.startswith("0800"):
            raise ValueError(f"Phone must match format {FOOTER_PHONE_FORMAT_EXAMPLE_0800}.")
        return _format_footer_phone_0800_from_digits(digits)
    if len(digits) != FOOTER_PHONE_DIGITS_COUNT:
        raise ValueError(f"Phone must match format {FOOTER_PHONE_FORMAT_EXAMPLE_MOBILE}.")
    return _format_footer_phone_mobile_from_digits(digits)


def format_footer_phone(value: str, phone_format: str = FOOTER_PHONE_FORMAT_MOBILE) -> str:
    normalized_format = _normalize_phone_format(phone_format)
    digits = _normalize_footer_phone_digits(value, normalized_format)
    if normalized_format == FOOTER_PHONE_FORMAT_TOLL_FREE_0800:
        if len(digits) == FOOTER_PHONE_DIGITS_COUNT and digits.startswith("0800"):
            return _format_footer_phone_0800_from_digits(digits)
        return str(value or "").strip()
    if len(digits) == FOOTER_PHONE_DIGITS_COUNT:
        return _format_footer_phone_mobile_from_digits(digits)
    return str(value or "").strip()


def _normalize_phone_format(phone_format: str) -> str:
    normalized = str(phone_format or "").strip().lower()
    if normalized == FOOTER_PHONE_FORMAT_TOLL_FREE_0800:
        return FOOTER_PHONE_FORMAT_TOLL_FREE_0800
    return FOOTER_PHONE_FORMAT_MOBILE


def _normalize_footer_phone_digits(value: str, phone_format: str) -> str:
    digits = re.sub(r"\D+", "", str(value or ""))
    if not digits:
        return ""
    if phone_format == FOOTER_PHONE_FORMAT_TOLL_FREE_0800:
        return _normalize_0800_digits(digits)
    if digits.startswith("380"):
        return digits[2:12]
    if digits.startswith("38"):
        rest = digits[2:]
        if not rest:
            return ""
        if rest.startswith("0"):
            return rest[:FOOTER_PHONE_DIGITS_COUNT]
        return f"0{rest}"[:FOOTER_PHONE_DIGITS_COUNT]
    if digits.startswith("0"):
        return digits[:FOOTER_PHONE_DIGITS_COUNT]
    if len(digits) >= FOOTER_PHONE_DIGITS_COUNT:
        tail = digits[-FOOTER_PHONE_DIGITS_COUNT:]
        if tail.startswith("0"):
            return tail
        return f"0{tail}"[:FOOTER_PHONE_DIGITS_COUNT]
    return f"0{digits}"[:FOOTER_PHONE_DIGITS_COUNT]


def _normalize_0800_digits(digits: str) -> str:
    if digits.startswith("380800"):
        return f"0{digits[3:]}"[:FOOTER_PHONE_DIGITS_COUNT]
    if digits.startswith("800"):
        return f"0{digits}"[:FOOTER_PHONE_DIGITS_COUNT]
    if digits.startswith("0800"):
        return digits[:FOOTER_PHONE_DIGITS_COUNT]
    if len(digits) >= 6:
        return f"0800{digits[-6:]}"[:FOOTER_PHONE_DIGITS_COUNT]
    return f"0800{digits}"[:FOOTER_PHONE_DIGITS_COUNT]


def _format_footer_phone_mobile_from_digits(digits: str) -> str:
    operator = digits[:3]
    left = digits[3:6]
    middle = digits[6:8]
    right = digits[8:10]
    return f"38 ({operator}) {left}-{middle}-{right}"


def _format_footer_phone_0800_from_digits(digits: str) -> str:
    return f"{digits[0]} ({digits[1:4]}) {digits[4:7]}-{digits[7:10]}"
